### DISCUSSION TITLE
Prevent App_Tables default Search

### DISCUSSION THREAD

**Discussion by lewysigns** [anvilistas/tabulator#105] [Q&A] [answered]
Is there a way to prevent the automatic ```.search()``` call to be made when using app_tables in the table options?

I want to set the table to a ```client_readable``` table, but I want to control the query on the table instead of loading all the data at once.

**Comment by lewysigns** [comment]
I found this, reading the code:

```python
@tabulator_module("query")
class QueryModule(AbstractModule):
    def __init__(self, mod, table):
        super().__init__(mod, table)
        mod.registerTableOption("initialQuery", EMPTY_QUERY)
        mod.registerTableFunction("setQuery", self.set_query)
        mod.registerTableFunction("clearQuery", self.clear_query)
        self.query = EMPTY_QUERY

    def initialize(self):
        if self.table.options.get("appTable") is None:
            return
        self.mod.subscribe("data-params", self.query_params)
        self.query = self.table.options.get("initialQuery", EMPTY_QUERY)
```

So I *Think* I can use the ```initial_query``` option to achieve what I am trying to do.

**Reply by lewysigns** [reply]
I attempt to set the option as the following, and it seems to make all subsequent queries fail.

```python
      self.tabulator.options = {
        "app_table": anvil.server.call('get_complete_loads',self.date_picker_start.date,self.date_picker_end.date),
        "initial_query":{'column_name':q.any_of(*self.names)},
        "height" :400,
        "pagination": False,
        "loading_indicator": True, # defaults to True - whether to use the anvil spinner during page loading
        "progressive_load": "load", # enable progressive loading
        "progressive_load_delay": 100, # wait 200 milliseconds between each request
      }
```

**Reply by lewysigns** [reply]
Reworking the server call , I get this tabulator error, in the web console, when trying to set the ```initial_query```  


```
runner.js:433 Uncaught runtime error: Uncaught TypeError: Cannot read properties of undefined (reading 'then') at https://cdn.skypack.dev/-/tabulator-tables@v5.4.3-Vy9m0B6vzJzRwJ1amSyz/dist=es2020,mode=imports,min/optimized/tabulator-tables.js:17, column 18624 TypeError: Cannot read properties of undefined (reading 'then')
    at Ei.load (tabulator-tables.js:17:18624)
    at _._loadInitialData (tabulator-tables.js:17:37879)
    at _._create (tabulator-tables.js:17:36035)
    at tabulator-tables.js:17:34581
handleJsError @ runner.js:433
runner.js:309 CLIENT: app_log-WARNING 2023-04-12 14:46:00: <class 'ExternalError'>

tabulator-tables.js:17 Uncaught TypeError: Cannot read properties of undefined (reading 'then')
    at Ei.load (tabulator-tables.js:17:18624)
    at _._loadInitialData (tabulator-tables.js:17:37879)
    at _._create (tabulator-tables.js:17:36035)
    at tabulator-tables.js:17:34581
```

**Reply by s-cork** [reply]
Can you provide a minimal example clone so I can take a look. 

The initial_query option is new so I can imagine it's got some hard edges.

**Reply by lewysigns** [reply]
Here is the clone link:

https://anvil.works/build#clone:TX5IJADW5BLXJTRB=IGU6N4GJMVJGZXZHMS4NRCBH


Looking at the code, (and I'd like to state I don't follow it fully), it looks like the ```self.query``` is set to the actual dictionary I am passing into the ```initial_query``` option instead of passing those arguements into a ```QUERY``` object. But I also can't find where self.query is used, so I could be missing something.

This line:

```python
self.query = self.table.options.get("initialQuery", EMPTY_QUERY)
```

I have a fork that changes it to this:

```python
        query_args = self.table.options.get("initialQuery")
        self.query = Query(**query_args) if query_args else EMPTY_QUERY
```

Not sure how to test locally.  Let me know if I should submit a pull request or if I am not on the right track!

**Reply by s-cork** [reply, marked as answer]
Ah yeah you're right in the issue but the fix is wrong here because the point was to allow both args and kwargs. 

Sorry i should have pointed you to this discussion:

https://github.com/s-cork/Tabulator/issues/98

The correct fix is gonna be document and throw a TypeError if an instance of Query is not provided.

**Reply by s-cork** [reply]
Note we could absolutely support dictionaries as well we'd just need to check for instance of dict in the code

**Comment by lewysigns** [comment]
That's exactly what I was looking for.

I think with the Query object being accessible, it probably is better to keep it fixed to just using that.

Thank you Stu!
